File: backend/src/app/geoip/denials.py
"""Denial accounting for the GeoIP/CrowdSec gate — persistent, flood-proof.

Two tiers (DR-G9, migration 040):

- ``geoip_denial_stats`` — per (day, reason, country) aggregate, upserted by
  the flush job. Bounded by days × ~4 reasons × ~250 countries, it counts
  EVERY denial and is the source of truth for totals.
- ``geoip_denial_events`` — individual rows for the recent-denials table,
  hard-capped per flush interval (``_EVENTS_PER_FLUSH``): under a flood only
  a sample lands here while the aggregate still counts everything. Pruned by
  the flush job itself (batched, oldest-first, own ts index).

The request path only touches bounded in-memory buffers (a scanner cycling
random IPs can neither flood a table nor grow memory); a 15s scheduler job
flushes them with its own session. ``_totals``/``_total_countries`` mirror
the all-time sums (hydrated from the stats table at startup, incremented on
record) so the Prometheus counters stay monotonic ACROSS restarts without a
DB query per scrape.
"""

from __future__ import annotations

from collections import Counter, deque
from datetime import UTC, datetime, timedelta

import structlog
from sqlalchemy import delete, func, select, text
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.models import GeoipDenialEvent, GeoipDenialStat
from app.geoip import lookup

log = structlog.get_logger("app.geoip")

_EVENTS_PER_FLUSH = 50  # row-sample cap per flush interval; aggregate counts all
_EVENT_RETENTION_DAYS = 30
_STATS_RETENTION_DAYS = 365
_PRUNE_BATCH = 10_000

_FAIL_OPEN_REASON = "fail_open"  # stats row for DR-G5 allows; not a denial

# --- request-path buffers (bounded) -------------------------------------------
_pending_events: deque[dict] = deque(maxlen=_EVENTS_PER_FLUSH)
_pending_agg: Counter[tuple[str, str]] = Counter()  # (reason, country) -> n

# All-time mirrors for Prometheus (hydrated + incremented; monotonic).
_totals: Counter[str] = Counter()
_total_countries: Counter[str] = Counter()
_total_fail_open = 0
_hydrated = False


def record(ip: str | None, country: str | None, path: str, reason: str) -> None:
    _pending_agg[(reason, country or "??")] += 1
    _totals[reason] += 1
    _total_countries[country or "??"] += 1
    _pending_events.append(
        {
            "ts": datetime.now(UTC),
            "ip": (ip or "?")[:45],
            "country": country,
            "path": path[:255],
            "reason": reason[:32],
        }
    )


def record_fail_open() -> None:
    global _total_fail_open
    _pending_agg[(_FAIL_OPEN_REASON, "??")] += 1
    _total_fail_open += 1


def prometheus_series() -> tuple[dict[str, int], dict[str, int], int]:
    """All-time (by_reason, by_country, fail_open) — monotonic across restarts."""
    return dict(_totals), dict(_total_countries), _total_fail_open


async def hydrate(session: AsyncSession) -> None:
    """Load all-time sums from the stats table into the Prometheus mirrors.

    Runs once at startup (before traffic); without it the counters would
    restart at 0 and long-term dashboards would see false resets."""
    global _hydrated, _total_fail_open
    rows = (
        await session.execute(
            select(
                GeoipDenialStat.reason,
                GeoipDenialStat.country,
                func.sum(GeoipDenialStat.count),
            ).group_by(GeoipDenialStat.reason, GeoipDenialStat.country)
        )
    ).all()
    for reason, country, total in rows:
        n = int(total or 0)
        if reason == _FAIL_OPEN_REASON:
            _total_fail_open += n
            continue
        _totals[reason] += n
        _total_countries[country] += n
    _hydrated = True
    log.info("geoip.denials_hydrated", reasons=dict(_totals), fail_open=_total_fail_open)


async def flush(session: AsyncSession) -> int:
    """Write pending buffers to the DB. Returns the number of aggregated denials.

    Caller (scheduler job) owns the commit. Buffers are swapped out first so a
    failed commit loses at most one interval — acceptable for statistics,
    and the Prometheus mirrors are incremented at record time regardless.
    """
    if not _pending_agg and not _pending_events:
        return 0
    agg = dict(_pending_agg)
    _pending_agg.clear()
    events = list(_pending_events)
    _pending_events.clear()

    bucket = datetime.now(UTC).date()
    for (reason, country), n in agg.items():
        # MariaDB-native upsert (repo rule: no ON CONFLICT idioms).
        await session.execute(
            text(
                "INSERT INTO geoip_denial_stats (bucket, reason, country, count) "
                "VALUES (:bucket, :reason, :country, :n) "
                "ON DUPLICATE KEY UPDATE count = count + :n"
            ),
            {"bucket": bucket, "reason": reason, "country": country, "n": n},
        )
    for e in events:
        session.add(GeoipDenialEvent(**e))
    await session.flush()
    return sum(agg.values())


async def prune(session: AsyncSession) -> None:
    """Retention: batched oldest-first deletes (unbounded DELETE gap-locks)."""
    event_cutoff = datetime.now(UTC) - timedelta(days=_EVENT_RETENTION_DAYS)
    while True:
        result = await session.execute(
            delete(GeoipDenialEvent).where(
                GeoipDenialEvent.id.in_(
                    select(GeoipDenialEvent.id)
                    .where(GeoipDenialEvent.ts < event_cutoff)
                    .order_by(GeoipDenialEvent.ts)
                    .limit(_PRUNE_BATCH)
                    .scalar_subquery()
                )
            )
        )
        await session.commit()
        if (result.rowcount or 0) < _PRUNE_BATCH:
            break
    stats_cutoff = (datetime.now(UTC) - timedelta(days=_STATS_RETENTION_DAYS)).date()
    await session.execute(delete(GeoipDenialStat).where(GeoipDenialStat.bucket < stats_cutoff))
    await session.commit()


async def snapshot(session: AsyncSession, limit: int = 50) -> dict:
    """Persistent view for the Access page (stats table + recent events).

    Pending not-yet-flushed denials are folded in so the UI never lags the
    15s flush interval on the counters."""
    rows = (
        await session.execute(
            select(
                GeoipDenialStat.reason,
                GeoipDenialStat.country,
                func.sum(GeoipDenialStat.count),
            ).group_by(GeoipDenialStat.reason, GeoipDenialStat.country)
        )
    ).all()
    by_reason: Counter[str] = Counter()
    by_country: Counter[str] = Counter()
    fail_open = 0
    for reason, country, total in rows:
        n = int(total or 0)
        if reason == _FAIL_OPEN_REASON:
            fail_open += n
            continue
        by_reason[reason] += n
        by_country[country] += n
    for (reason, country), n in _pending_agg.items():
        if reason == _FAIL_OPEN_REASON:
            fail_open += n
        else:
            by_reason[reason] += n
            by_country[country] += n

    since = await session.scalar(select(func.min(GeoipDenialStat.bucket)))
    recent_rows = (
        (
            await session.execute(
                select(GeoipDenialEvent).order_by(GeoipDenialEvent.ts.desc()).limit(limit)
            )
        )
        .scalars()
        .all()
    )

    def _hover_name(ip: str, code: str | None) -> str | None:
        # Display-only hover label; attached only while the live DB still
        # agrees with the code stored at denial time.
        live_code, name = lookup.country_display(ip)
        return name if live_code == code else None

    pending = [
        {
            "at": e["ts"].isoformat(),
            "ip": e["ip"],
            "country": e["country"],
            "country_name": _hover_name(e["ip"], e["country"]),
            "path": e["path"],
            "reason": e["reason"],
        }
        for e in reversed(_pending_events)
    ]
    persisted = [
        {
            "at": e.ts.isoformat(),
            "ip": e.ip,
            "country": e.country,
            "country_name": _hover_name(e.ip, e.country),
            "path": e.path,
            "reason": e.reason,
        }
        for e in recent_rows
    ]
    return {
        "since": since.isoformat() if since else None,
        "total": sum(by_reason.values()),
        "by_reason": dict(by_reason),
        "top_countries": [{"country": c, "count": n} for c, n in by_country.most_common(20)],
        "fail_open_allows": fail_open,
        "recent": (pending + persisted)[:limit],
    }


async def flush_job() -> None:
    """Scheduler entry point (15s): flush buffers with an own session."""
    from app.db.base import get_sessionmaker  # local import: avoid cycles

    if not _pending_agg and not _pending_events:
        return
    async with get_sessionmaker()() as session:
        n = await flush(session)
        await session.commit()
    if n:
        log.debug("geoip.denials_flushed", denials=n)


async def prune_job() -> None:
    """Scheduler entry point (daily): retention for events + stats."""
    from app.db.base import get_sessionmaker

    async with get_sessionmaker()() as session:
        await prune(session)

File: orbit/lib/orbit/geoip/denials.ex
defmodule Orbit.GeoIP.Denials do
  @moduledoc """
  Denial accounting for the GeoIP/CrowdSec gate (DR-G9) — port of
  geoip/denials.py, writing the SAME alembic-owned tables as the python
  stack (upserts are additive, prune is idempotent — both stacks can run).

  Two tiers:
  - `geoip_denial_stats` — per (day, reason, country) aggregate, upserted
    on flush. Bounded by days × ~4 reasons × ~250 countries; counts EVERY
    denial and is the source of truth for totals.
  - `geoip_denial_events` — individual rows for the recent-denials view,
    hard-capped per flush interval (#{50}): under a flood only a sample
    lands here while the aggregate still counts everything.

  The request path only casts into bounded in-process buffers (a scanner
  cycling random IPs can neither flood a table nor grow memory); a 15s
  timer flushes, a daily timer prunes (batched oldest-first deletes — an
  unbounded DELETE gap-locks, repo incident rule). Buffers are swapped out
  before the write, so a failed flush loses at most one interval —
  acceptable for statistics.

  The python-side Prometheus mirrors / snapshot view stay python-owned
  until orbit grows its Access surface; this module's job is that no orbit
  denial is lost to those surfaces.
  """

  use GenServer

  require Logger

  @flush_ms 15_000
  @prune_ms 24 * 60 * 60_000
  @events_per_flush 50
  @event_retention_days 30
  @stats_retention_days 365
  @prune_batch 10_000

  @fail_open_reason "fail_open"

  def start_link(opts) do
    GenServer.start_link(__MODULE__, :ok, name: Keyword.get(opts, :name, __MODULE__))
  end

  @doc "Count one denial (bounded buffers; safe without the server: drop)."
  def record(server \\ __MODULE__, ip, country, path, reason) do
    GenServer.cast(server, {:record, ip, country, path, reason})
  catch
    :exit, _ -> :ok
  end

  @doc "Count one DR-G5 fail-open allow (stats row only, not a denial)."
  def record_fail_open(server \\ __MODULE__) do
    GenServer.cast(server, :record_fail_open)
  catch
    :exit, _ -> :ok
  end

  @doc "Synchronous flush (test/ops seam); returns the aggregated count."
  def flush(server \\ __MODULE__), do: GenServer.call(server, :flush)

  # -- pure buffer transitions (unit-tested) --------------------------------

  @doc false
  def empty_buffers, do: %{agg: %{}, events: []}

  @doc false
  def add_denial(buffers, ip, country, path, reason, now) do
    event = %{
      ts: now,
      ip: String.slice(ip || "?", 0, 45),
      country: country,
      path: String.slice(path, 0, 255),
      reason: String.slice(reason, 0, 32)
    }

    %{
      agg: Map.update(buffers.agg, {reason, country || "??"}, 1, &(&1 + 1)),
      # Newest-first, hard cap: python's deque(maxlen) keeps the newest
      # sample under a flood; prepend + take does the same here.
      events: Enum.take([event | buffers.events], @events_per_flush)
    }
  end

  @doc false
  def add_fail_open(buffers) do
    %{buffers | agg: Map.update(buffers.agg, {@fail_open_reason, "??"}, 1, &(&1 + 1))}
  end

  # -- GenServer ------------------------------------------------------------

  @impl true
  def init(:ok) do
    Process.send_after(self(), :flush, @flush_ms)
    Process.send_after(self(), :prune, @prune_ms)
    # Lock-free counter for the all-pages footer. An ETS read costs
    # microseconds and needs no message to this process — a GenServer.call
    # on every rendered page would serialise the whole site through here,
    # and a COUNT(*) per page would hit the database just as often.
    :ets.new(@counter_table, [:named_table, :public, :set, write_concurrency: true])
    :ets.insert(@counter_table, {:blocked, 0})
    {:ok, empty_buffers()}
  end

  @counter_table :orbit_geoip_denial_count

  @doc """
  Denials since this process started (footer badge).

  Deliberately process-local and not a query: it answers "is the gate doing
  anything right now", the same live-health semantics as the hub counters.
  Returns 0 before the table exists (boot, or tests without the supervisor).
  """
  def blocked_count do
    case :ets.lookup(@counter_table, :blocked) do
      [{:blocked, n}] -> n
      _ -> 0
    end
  rescue
    ArgumentError -> 0
  end

  @impl true
  def handle_cast({:record, ip, country, path, reason}, buffers) do
    bump_counter()
    {:noreply, add_denial(buffers, ip, country, path, reason, DateTime.utc_now())}
  end

  defp bump_counter do
    :ets.update_counter(@counter_table, :blocked, 1)
  rescue
    ArgumentError -> 0
  end

  def handle_cast(:record_fail_open, buffers) do
    {:noreply, add_fail_open(buffers)}
  end

  @impl true
  def handle_call(:flush, _from, buffers) do
    {n, rest} = do_flush(buffers)
    {:reply, n, rest}
  end

  @impl true
  def handle_info(:flush, buffers) do
    Process.send_after(self(), :flush, @flush_ms)
    {_n, rest} = do_flush(buffers)
    {:noreply, rest}
  end

  def handle_info(:prune, buffers) do
    Process.send_after(self(), :prune, @prune_ms)
    prune()
    {:noreply, buffers}
  end

  # Swap-then-write: a failed flush loses at most one interval (moduledoc).
  defp do_flush(%{agg: agg, events: events}) when agg == %{} and events == [] do
    {0, empty_buffers()}
  end

  defp do_flush(%{agg: agg, events: events}) do
    bucket = Date.utc_today()

    for {{reason, country}, n} <- agg do
      # MariaDB-native upsert (repo rule: no ON CONFLICT idioms).
      Orbit.Repo.query!(
        "INSERT INTO geoip_denial_stats (bucket, reason, country, count) " <>
          "VALUES (?, ?, ?, ?) ON DUPLICATE KEY UPDATE count = count + ?",
        [bucket, reason, country, n, n]
      )
    end

    for e <- Enum.reverse(events) do
      Orbit.Repo.query!(
        "INSERT INTO geoip_denial_events (ts, ip, country, path, reason) VALUES (?, ?, ?, ?, ?)",
        [DateTime.to_naive(e.ts), e.ip, e.country, e.path, e.reason]
      )
    end

    n = agg |> Map.values() |> Enum.sum()
    if n > 0, do: Logger.debug("geoip.denials_flushed denials=#{n}")
    {n, empty_buffers()}
  rescue
    error ->
      Logger.warning("geoip.denials_flush_failed error=#{Exception.message(error)}")
      {0, empty_buffers()}
  end

  defp prune do
    event_cutoff =
      DateTime.utc_now() |> DateTime.add(-@event_retention_days, :day) |> DateTime.to_naive()

    prune_events_batched(event_cutoff)

    stats_cutoff = Date.add(Date.utc_today(), -@stats_retention_days)
    Orbit.Repo.query!("DELETE FROM geoip_denial_stats WHERE bucket < ?", [stats_cutoff])
    :ok
  rescue
    error -> Logger.warning("geoip.denials_prune_failed error=#{Exception.message(error)}")
  end

  defp prune_events_batched(cutoff) do
    %{num_rows: n} =
      Orbit.Repo.query!(
        "DELETE FROM geoip_denial_events WHERE ts < ? ORDER BY ts LIMIT ?",
        [cutoff, @prune_batch]
      )

    if n >= @prune_batch, do: prune_events_batched(cutoff)
  end
end
